cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,377 @@
+        bool is_valid = thrust::get<0>(null_and_null_count);
+        T factor      = thrust::get<1>(null_and_null_count);
+
+        T ci = pair.first;
+        T cj = pair.second;
+
+        if (is_valid and (factor != 0.0)) {
+          return {ci / factor, cj / factor};
+        } else {
+          return {ci, cj};
+        }
+      });
+  }
+  compute_recurrence(pairs, stream);
+  // copy the second elements to the output for now
+  thrust::transform(rmm::exec_policy(stream),
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin(),
+                    [=] __device__(pair_type pair) -> T { return pair.second; });
+  return output;
+}
+
+template <typename T>
+std::unique_ptr<column> ewma(std::unique_ptr<aggregation> const& agg,
+                             column_view const& input,
+                             rmm::cuda_stream_view stream,
+                             rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(cudf::is_floating_point(input.type()), "Column must be floating point type");
+  bool adjust = (dynamic_cast<ewma_aggregation*>(agg.get()))->adjust;
+  T com       = (dynamic_cast<ewma_aggregation*>(agg.get()))->com;
+  // center of mass is easier for the user, but the recurrences are
+  // better expressed in terms of the derived parameter `beta`
+  T beta = 1.0 - (1.0 / (com + 1.0));
+
+  rmm::device_uvector<T> data(input.size(), stream, mr);
+  if (adjust) {
+    data = compute_ewma_adjust(input, beta, stream, mr);
+  } else {
+    data = compute_ewma_noadjust(input, beta, stream, mr);
+  }
+  auto col = std::make_unique<column>(
+    cudf::data_type(cudf::type_to_id<T>()), input.size(), std::move(data.release()));
+  return col;
+}
+
+struct ewma_functor {
+  template <typename T>
+  std::unique_ptr<column> operator()(std::unique_ptr<aggregation> const& agg,
+                                     column_view const& input,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    return std::unique_ptr<column>();
+  }
+};
+
+template <>
+std::unique_ptr<column> ewma_functor::operator()<float>(std::unique_ptr<aggregation> const& agg,
+                                                        column_view const& input,
+                                                        rmm::cuda_stream_view stream,
+                                                        rmm::mr::device_memory_resource* mr)
+{
+  return ewma<float>(agg, input, stream, mr);
+}
+
+template <>
+std::unique_ptr<column> ewma_functor::operator()<double>(std::unique_ptr<aggregation> const& agg,
+                                                         column_view const& input,
+                                                         rmm::cuda_stream_view stream,
+                                                         rmm::mr::device_memory_resource* mr)
+{
+  return ewma<double>(agg, input, stream, mr);
+}
```

Comment:
FYI you can put the specializations directly into the class rather than defining them out of band like this. However, I'm more confused about why the default is to return a default-initialized unique pointer to a column. Should that be an error case?